### PR TITLE
C++: Fix FormattingFunction regression.

### PR DESCRIPTION
**Pull request by geoffw0**
Fix the `FormattingFunction` regression caused by https://github.com/github/codeql/pull/5059, and identified in the [CPP Differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1691/) job yesterday.

---

Investigation:

boostorg/boost
 - 4 results for `cpp/wrong-type-format-argument`, all look correct but were removed by https://github.com/github/codeql/pull/5059
 - 1 result for `cpp/non-constant-format`, looks correct but removed by https://github.com/github/codeql/pull/5059
 - 1 result for `cpp/path-injection` on the diff that I can't reproduce locally
 - all of the results are on calls to a function `report_error`, which calls `vsnprintf`.
 - this snapshot has two implementations of `vsnprintf`, one has an implementation outside the source directory.

torvalds/linux
 - 10 results for `cpp/wrong-type-format-argument`, all were removed by https://github.com/github/codeql/pull/5059
 - 14 results for `cpp/non-constant-format`, not affected by the change
 - 23 results for `cpp/path-injection`, not affected by the change
 - 37 results for `cpp/overrunning-write`, all were removed by https://github.com/github/codeql/pull/5059
 - a sample of the lost results include calls to `die` (which calls `vfprintf`) and `sprintf` (which calls `vsnprintf`).
 - this snapshot has one implementation of `vfprintf`, that has  an implementation outside the source directory.
 - this snapshot has three implementations of `vsnprintf`, one has an implementation outside the source directory, one has only a declaration that is inside the source directory, and the last has an implementation inside the source directory (problematic).

My previous experiments suggested that the `exists(f.getBlock())` (having source not just a declaration) and `exists(f.getFile().getRelativePath())` (being outside the source directory) solutions both excluded the results we wanted to exclude with slightly different tradeoffs, so requiring both conditions to exclude results ought to preserve the changes we want while making the exclusion more precise.  In the case of Linux, one of the versions of `vsnprintf` (marked 'problematic' above) will still be excluded and we'll still lose results for it (all of the `cpp/overrunning-write` results as it happens).  Looking at the function in question I don't think there's any getting around this if we want to exclude custom printf variants, because it is one:
```
 * This function generally follows C99 vsnprintf, but has some
 * extensions and a few limitations:
...
```

---

TL;DR: this improvement should mostly, but not entirely, restore the cpp-differences we saw yesterday.  **Please wait for cpp-differences on this branch to confirm everything is as expected before merging.**

### CODE REVIEW

**Comment by jbj** [comment]
This looks good, but my confidence is a bit shaken by your findings about printf functions with multiple declarations and possibly multiple definitions. It makes me wonder if we should somehow require that it's the _same_ `FunctionDeclarationEntry` that's in the snapshot and has a body. That's at least what I had in mind when I thought about it informally.

If CPP-Differences shows a clear improvement, I'll merge this PR as it is.

**Comment by geoffw0** [comment]
> This looks good, but my confidence is a bit shaken by your findings about printf functions with multiple declarations and possibly multiple definitions.

Sorry I wasn't clear.  I doubt they're actually the same functions, I just did a query for all functions with the name `"vsnprintf"` (etc) to see what was out there.  I have no reason to believe any one particular call isn't associated with one well-defined function, I just didn't check which in all cases.

I'll have a diff running soon.

**Comment by geoffw0** [comment]
Differences job: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1698/ (if I've set it up correctly it has the same 'before' as https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1691/, and the same after but with the commit from this PR bolted on as well.  So we're hoping to see a much smaller set of changes)

**Review by jbj — APPROVED** [review]
The CPP-Differences LGTM. It's what we expected: Linux and Boost lose results because they have their own printf implementations.

**Comment by geoffw0** [comment]
> It's what we expected: Linux and Boost lose results because they have their own printf implementations.

Yes, Linux still loses some results because of its own printf implementation.  But Boost appears to (still) lose a single result but that only seems to depend on `SnprintfImpl`, not `UserDefinedFormattingFunction` / `predicate primitiveVariadicFormatter(
`, so I'm not sure what's going on there.

**Comment by geoffw0** [comment]
The lost result on Boost (`cpp/path-injection`) was caused by https://github.com/github/codeql/pull/5065 (@MathiasVP), which happened to be in the same day's diff as this PR.  I'm not sure exactly what the data flow is but there's a call to `sscanf` in the same function, so this is plausibly a correct change.

**Comment by MathiasVP** [comment]
> The lost result on Boost (`cpp/path-injection`) was caused by #5065 (@MathiasVP), which happened to be in the same day's diff as this PR. I'm not sure exactly what the data flow is but there's a call to `sscanf` in the same function, so this is plausibly a correct change.

I can't reproduce losing that result when running the latest `main` (8a2e063af7c48dc622ab07b0d7f5dc1aa8761516) locally. (It's a result we gained after https://github.com/github/codeql/pull/4302.) Can you reproduce the CPP-difference behavior locally?

**Comment by geoffw0** [comment]
Yes, I reproduced it by checking out the submodule commits of the before and after commits of https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1691/ (i.e. 749dfe4358425e7beb8704ca3e6588da23ab0fec and 63362a52c5daec7283b9a1f574f4df38f03b9e9c).

**Comment by MathiasVP** [comment]
Oh, I see. The result is restored by 7087904637a30be28fc6ffac17beb85cc20e666c.

**Comment by MathiasVP** [comment]
Hm! The result disappears when we have both https://github.com/github/codeql/pull/4784 and https://github.com/github/codeql/pull/5065. Weird. https://github.com/github/codeql/pull/4784/commits/fd8034cd8ccfd13d34d10d9b38b2312603d78361 does introduce a peculiar dataflow rule that adds flow into functions we don't have a dataflow model flow. It shouldn't affect anything, though, as https://github.com/github/codeql/pull/5065 only adds taintflow models.

Anyway, https://github.com/github/codeql/pull/4784 has been reverted, which is why we have the flow on 7087904637a30be28fc6ffac17beb85cc20e666c, but not on 63362a52c5daec7283b9a1f574f4df38f03b9e9c.
